opensquirrel/passes/exporter/cqasmv1_exporter.py

```diff
@@ -62,15 +65,17 @@ def visit_gate(self, gate: Gate) -> None:
         if any(not isinstance(arg, Qubit) for arg in gate.arguments):  # type: ignore[union-attr]
             params = [arg.accept(self) for arg in gate.arguments if not isinstance(arg, Qubit)]  # type: ignore[union-attr]
         qubit_args = (arg.accept(self) for arg in gate.arguments if isinstance(arg, Qubit))  # type: ignore[union-attr]
-        self.cqasmv1_string += "{} {}{}\n".format(
-            gate_name, ", ".join(qubit_args), ", " + ", ".join(params) if params else ""
-        )
+        self.output += "{} {}{}\n".format(gate_name, ", ".join(qubit_args), ", " + ", ".join(params) if params else "")
 
 
 def export(circuit: Circuit) -> str:
-    cqasmv1_creator = _CQASMv1Creator(circuit.register_manager)
-
+    barrier_links = [
+        list(barrier_link)
+        for barrier, barrier_link in groupby(circuit.ir.statements, lambda stmt: isinstance(stmt, Barrier))
+        if barrier
+    ]
+    cqasmv1_creator = _CQASMv1Creator(circuit.register_manager, barrier_links)
```

Comment:
Why don't we?:
- Keep the visitor simple. For example, when a barrier comes, it will just print it.
- So we do not pass any barrier links to the visitor.
- We cannot just copy the IR and transform it to convert each group of 'linked' barriers into a list, because the visitor interface doesn't understand about nodes that are list of barriers.
- What we can do is a simple post processing (yes, maybe slower, because it's an extra pass, but clearer) which goes over the final program and transforms something like `barrier q[i]\nbarrier q[j]` into `barrier q[i, j]`:

```
def dump_barrier_group(indices: list[int]) -> str:
    return "barrier q[{}]".format(", ".join(indices))

def merge_barrier_groups(output: str) -> str:
    new_output = ""
    barrier_group_indices = []
    for line in output.split('\n'):
        if not line.begin("barrier"):
            new_output += dump_barrier_group(barrier_group_indices)
            new_output += line
            barrier_group_indices = []
        else:
            barrier_group_indices.extend(get_barrier_indices(line))
    if len(barrier_group_indices) != 0:
        new_output += dump_barrier_group(current_barrier_group_indices)
```

Then:

```
    output = cqasmv1_creator.output.rstrip() + "\n"
    return merge_barrier_groups(output)
```

Reply:
I didn't want to do it like this, because then the only criteria were using to determine whether something is a barrier is if the string is equal to "barrier". Doesn't _feel_ robust (but I must admit, it is more just a _feeling_). I do like to keep the visitor class clean though, so I'd be happy to implement it like this. What do you think?